pyexchange/bitso.py

```diff
@@ -0,0 +1,339 @@
+                                    timestamp=self._iso8601_to_unix(item['created_at']),
+                                    pair=item['book'],
+                                    is_sell=item['maker_side'] == 'buy',
+                                    price=Wad.from_number(item['price']),
+                                    amount=Wad.from_number(abs(float(item['amount'])))), result))
+
+    def _http_request(self, method: str, resource: str, params: dict):
+        assert(isinstance(method, str))
+        assert(isinstance(resource, str))
+        assert(isinstance(params, dict) or (params is None))
+
+        url=f"{self.api_server}{resource}"
+        if params:
+            url=f"{self.api_server}{resource}?{urlencode(params)}"
+
+        return self._result(requests.request(method=method,
+                                             url=url,
+                                             timeout=self.timeout))
+
+    # Interprets the response to an HTTP GET, POST or DELETE request
+    def _http_authenticated_request(self, method: str, resource: str, params: dict, data: dict = {}):
+        assert(isinstance(method, str))
+        assert(isinstance(resource, str))
+        assert(isinstance(params, dict) or (params is None))
+        assert(isinstance(data, dict))
+
+        nonce = str(int(round(time.time() * 1000)))
+
+        # if has params, else strip out query params, otherwise call fails
+        if not params:
+            message = f'{nonce}{method}{resource}'
+            url = f"{self.api_server}{resource}"
+        else:
+            message = f'{nonce}{method}{resource}?{urlencode(params)}'
+            url = f"{self.api_server}{resource}?{urlencode(params)}"
+
+        if (method == "POST"):
+            message += json.dumps(data)
+
+        signature = hmac.new(self.secret_key.encode('utf-8'),
+                                            message.encode('utf-8'),
+                                            hashlib.sha256).hexdigest()
+        # Build the auth header
+        auth_header = 'Bitso %s:%s:%s' % (self.api_key, nonce, signature)
+
+        headers = {
+            "Authorization": auth_header
+        }
+
+        if method != "POST":
+            return self._result(requests.request(method=method,
+                                             url=url,
+                                             headers=headers,
+                                             timeout=self.timeout))
+
+        else:
+            return self._result(requests.request(method=method,
+                                             url=url,
+                                             headers=headers,
+                                             json=data,
+                                             timeout=self.timeout))
+    @staticmethod
+    def _result(result) -> dict:
+
+        if not result.ok:
+            raise Exception(f"Bitso API invalid HTTP response: {http_response_summary(result)}")
```

Comment:
Raising the base `Exception` class is frowned upon because it makes it more difficult for calling code to catch only expected exceptions.  I know the other exchanges in the library do this, but maybe we should break consistency rather than accumulate technical debt.

Reply:
Replaced with ValueError, as discussed.